src/kernel/terminal.c

```diff
@@ -28,129 +27,133 @@ void init_terminal(void)
  */
 void clear_terminal(void)
 {
-	unsigned int i = 0;
-
-	/* this loops clears the screen
-	 * there are 25 lines each of 80 columns;
-	 * each element takes 2 bytes */
-	while(i < SCREENSIZE)
+	for (int i = 0; i < NB_LINES; i++)
 	{
-		terminal.vidptr[i++] = '\0';
-		terminal.vidptr[i++] = LIGHT_GRAY;
+		clear_line(i);
 	}
 }
 
 /**
- * Add a newline on the terminal ('\n')
+ * Blank one line
  *
  * @return void
  */
-void newline_on_terminal(void)
+void	clear_line(int line_number)
 {
-	unsigned int line_size = BYTES_FOR_ELEMENT * NB_COLUMNS;
-	terminal.current_loc += line_size - terminal.current_loc % line_size;
-
-	move_cursor(terminal.current_loc / 2);
+	for (int i = 0;i < NB_COLUMNS; i++)
+	{
+		terminal.vidptr[line_number * NB_COLUMNS + i] = 0 | (LIGHT_GRAY << 8);
+	}
 }
 
 /**
- * Delete a char on the terminal ('\b')
+ * Add a newline on the terminal ('\n')
  *
  * @return void
  */
-void delete_on_terminal(void)
+void newline_on_terminal(void)
 {
-	if (terminal.current_loc - 2 >= 0)
+	if (terminal.current_loc / NB_COLUMNS == NB_LINES - 1)
+		scroll_down();
+	int filling = NB_COLUMNS - (terminal.current_loc % NB_COLUMNS);
+	if ((terminal.vidptr[terminal.current_loc] & 0xff) != '\0')
 	{
-		terminal.vidptr[terminal.current_loc - 2] = '\0';
-		terminal.current_loc -= 2;
-
-		move_cursor(terminal.current_loc / 2);
+		memcpy(&terminal.vidptr[terminal.current_loc + filling],
+			&terminal.vidptr[terminal.current_loc],
+			(SCREEN_SIZE - terminal.current_loc - filling) * sizeof(short));
 	}
-
-	move_buffer_terminal_to_left();
+	for (int i = 0; i < filling; i++)
+		terminal.vidptr[terminal.current_loc + i] = ' ' | (LIGHT_GRAY << 8);
+	terminal.current_loc += filling;
+	terminal.deletable = terminal.current_loc;
+	move_cursor(terminal.current_loc);
 }
 
 /**
- * Add a tab on the terminal ('\t')
+ * Delete a char on the terminal ('\b')
  *
  * @return void
  */
-void tab_on_terminal(void)
+void delete_on_terminal(void)
 {
-	int i = 0;
-	while (i++ < TAB_SIZE)
-		print_char_on_terminal(' ', LIGHT_GRAY);
+	if (terminal.current_loc <= terminal.deletable)
+		return;
+	if (terminal.current_loc > 0 && terminal.current_loc <= SCREEN_SIZE)
+	{
+		memcpy(&terminal.vidptr[terminal.current_loc - 1],
+			&terminal.vidptr[terminal.current_loc],
+			sizeof(short) * (SCREEN_SIZE - terminal.current_loc -1));
+		terminal.current_loc--;
+		move_cursor(terminal.current_loc);
+	}
 }
 
 /**
- * Move the buffer to the left of one char on deletion of a char
+ * Add a tab on the terminal ('\t')
  *
  * @return void
  */
-void move_buffer_terminal_to_left(void)
+void tab_on_terminal(char colour)
 {
-	int i = terminal.current_loc;
-
-	while (terminal.vidptr[i + 2] != '\0')
-	{
-		terminal.vidptr[i] = terminal.vidptr[i + 2];
-		terminal.vidptr[i + 1] = terminal.vidptr[i + 3];
-		i += 2;
-	}
-	terminal.vidptr[i] = '\0';
+	int i = 0;
+	short space = ' ' | (colour << 8);
+	if (terminal.current_loc + 4 >= SCREEN_SIZE)
```

Comment:
4 should not be a hardcoded value, use TAB_SIZE instead
```suggestion
	if (terminal.current_loc + TAB_SIZE >= SCREEN_SIZE)
```